File: examples/assume-contracts/src/main/java/example/repository/UserService.java
package example.repository;

/**
 * A small service over {@link UserRepository}. Its correctness is COMPOSITIONAL: it holds only if the
 * repository upholds an output property (a non-null user has a non-negative age, and {@code findById(id)}
 * returns a user whose {@code id} is {@code id}). The proofs supply that property with
 * {@code Bmc.assumeEvery(repo::findById, …)} — assume-guarantee — instead of a model of the repository.
 */
public final class UserService {

    private final UserRepository repo;

    public UserService(UserRepository repo) {
        this.repo = repo;
    }

    /**
     * The user's age, or {@code -1} when absent. Under the assumption that every {@code findById}
     * returns {@code null} or a user with {@code age >= 0}, the result is always {@code >= -1}.
     */
    public int ageOrAbsent(int id) {
        UserRepository.User u = repo.findById(id);
        if (u == null) {
            return -1;
        }
        return u.age();
    }

    /**
     * The id of the looked-up user, or {@code -1} when absent. Under the args-aware assumption that
     * {@code findById(id)} returns {@code null} or a user whose {@code id == id}, this returns either
     * {@code -1} or exactly {@code id}.
     */
    public int idOf(int id) {
        UserRepository.User u = repo.findById(id);
        if (u == null) {
            return -1;
        }
        return u.id();
    }
}

File: examples/assume-contracts-kotlin/src/main/kotlin/example/repository/UserService.kt
package example.repository

/**
 * A small service over [UserRepository]. Its correctness is COMPOSITIONAL: it holds only if the
 * repository upholds an output property (a non-null user has a non-negative age, and findById(id)
 * returns a user whose id is id). The proofs supply that property with
 * `Bmc.assumeEvery(repo::findById) { ... }` - assume-guarantee - instead of a model of the repository.
 */
class UserService(private val repo: UserRepository) {

    /**
     * The user's age, or -1 when absent. Under the assumption that every findById returns null or a
     * user with age >= 0, the result is always >= -1.
     */
    fun ageOrAbsent(id: Int): Int {
        val u = repo.findById(id) ?: return -1
        return u.age
    }

    /**
     * The id of the looked-up user, or -1 when absent. Under the args-aware assumption that
     * findById(id) returns null or a user whose id == id, this returns either -1 or exactly id.
     */
    fun idOf(id: Int): Int {
        val u = repo.findById(id) ?: return -1
        return u.id
    }
}
